Adds read-only property for the I2C EEPROM Target

Manu3l0us opened the pull request:
In case the EEPROM data stored in the buffer shall be not writable from the I2C interface, the read-only property can now be set in the device tree.

Commit:
i2c: target: eeprom_target: add read-only property

In case the EEPROM data stored in the buffer shall be not writable from the
I2C interface, the read-only property can now be set in the device tree.

Signed-off-by: Manuel Aebischer <manuel.aebischer@belden.com>

henrikbrixandersen commented on drivers/i2c/target/eeprom_target.c:
```diff
@@ -129,7 +130,7 @@ static int eeprom_target_write_received(struct i2c_target_config *config,
 		data->buffer_idx = val | (data->buffer_idx << 8);
 		data->idx_write_cnt++;
-	} else {
+	} else if (!data->read_only) {
```
This is insufficient. The I2C write transaction would need to be NAK'ed as well (see the comment just above the code, you've changed).

Manu3l0us replied:
Thanks for pointing it out, of course you are right

henrikbrixandersen reviewed (changes requested):
Looks good! Now all that's needed is a corresponding test case in `tests/drivers/i2c/i2c_target_api/` 👍

Manu3l0us commented:
> Looks good! Now all that's needed is a corresponding test case in `tests/drivers/i2c/i2c_target_api/` 👍

Good point, I will figure out how and update the PR once done.

Manu3l0us commented:
@henrikbrixandersen I am thinking about adding another EEPROM which is read-only to the device tree overlays within the test project. Also an EEPROM with 16bit addressing should be added to increase test coverage of the existing code which also supports this since a couple of months. I could probably do that at the same time. Does this sound like the right approach to you?

henrikbrixandersen commented:
> @henrikbrixandersen I am thinking about adding another EEPROM which is read-only to the device tree overlays within the test project. Also an EEPROM with 16bit addressing should be added to increase test coverage of the existing code which also supports this since a couple of months. I could probably do that at the same time. Does this sound like the right approach to you?

Yes, that sounds good along with some accompanying test cases.

Manu3l0us commented:
I took this up again as I was busy with other stuff during the last months. I tried to run the existing test on a RPi Pico but couldn't get it to work until now. What hardware setup is needed to run those tests?

Commit:
tests: i2c: i2c_target_api: Add test for read-only eeprom target

This additional test case verifies that an i2c_eeprom_target, which is
set as read-only in the device tree cannot be written from the I2C bus.

Signed-off-by: Manuel Aebischer <manuel.aebischer@belden.com>

Manu3l0us commented:
@henrikbrixandersen I finally had the time to write a simple test and I pushed my work in progress. Some questions:

1. On the i2c_dw based target I'm using here, no NACK is generated and therefore I don't get a non-zero return code. Therefore I currently commented this check out for the test to run. Should I add a platform dependent test condition or how would you handle this?
2. I added the overlay for the rpi_pico which I currently use and I updated the overlay for it. What is the general expectation when adding test cases? Shall I update the other overlays as well? I don't have the hardware to test any of those boards.

Manu3l0us commented:
Okay, I saw that the boards are tested in CI, so I can safely add it to the overlay without breaking the tests

henrikbrixandersen commented on tests/drivers/i2c/i2c_target_api/src/main.c:
```diff
@@ -273,4 +276,47 @@ ZTEST(i2c_eeprom_target, test_eeprom_target)
+	ret = i2c_write(i2c_1, write_attempt_data, TEST_DATA_SIZE, addr_2);
+
+	/* I2C_DW (and maybe others?) don't generate the NACK as expected */
+#ifndef CONFIG_I2C_DW
+	zassert_not_equal(ret, 0, "Write attempt to read-only EEPROM 2 succeeded unexpectedly");
+#endif
```
Why is this needed?

Manu3l0us replied:
Usually the target should generate a NAK if it's being written and it's write protected, therefore the write on the controller side should return a non-zero return code. Unfortunately, the DesignWare I2C does not seem to implement this NAKing when used in target mode. I see if I can get hold of some other boards (e.g. STM32) to verify this as well.

henrikbrixandersen reviewed (changes requested):
Thanks! However, I don't think we need this read-only EEPROM added to each and every board here (and we really need to cut down on these boards).

Perhaps make the test conditional on the device being available (skip it if not present) and just add it to the board(s), you've tested with?

Manu3l0us commented:
> Thanks! However, I don't think we need this read-only EEPROM added to each and every board here (and we really need to cut down on these boards).
> 
> Perhaps make the test conditional on the device being available (skip it if not present) and just add it to the board(s), you've tested with?

Okay, understood. Will clean this up the coming week hopefully.

decsny reviewed (commented):
dt prop is fine

msalau commented on drivers/i2c/target/eeprom_target.c:
```diff
@@ -249,34 +254,25 @@ static int i2c_eeprom_target_init(const struct device *dev)
+#define I2C_EEPROM_INIT(inst)                                                                      \
+	static struct i2c_eeprom_target_data i2c_eeprom_target_##inst##_dev_data = {               \
+		.address_width = DT_INST_PROP_OR(inst, address_width, 8),                          \
+		.read_only = DT_INST_PROP_OR(inst, read_only, 8),                                  \
```
Looks like a copy-paste error. The `read-only` property is always present due to its type being boolean so there is no reason to give it a default value.
```suggestion
		.read_only = DT_INST_PROP(inst, read_only),                                  \
```

henrikbrixandersen commented:
@Manu3l0us Any plans for resuming work on this?

henrikbrixandersen commented:
@Manu3l0us I'm going to close this due to lack of response, although I'd love to see this feature make it in. Please reopen it if you decide to come back to this.
